WillTDA/File-System-DB
# fix: typings and general cleanup

## Purpose

The intention of these changes was to add the `"types"` field to the `package.json` file to resolve #9.
However, I've also included general improvements to the structure of the project. Namely dropping the npm workspaces setup in favor of a more traditional repository structure. This should make it easier for contributors to understand the project and make changes.

## Changes

- Add `"name"` fields to jobs for GitHub Actions
- Add `"types"` field to `package.json` to resolve #9
- Add `// @ts-check` to `index.test.js` for static type checking (since it is not included in the `tsconfig.json`)
- Add configuration for `CODEOWNERS` to automatically add @WillTDA as a reviewer on new PRs (see [the GitHub docs](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) for more details on this)
- Add configuration for "Buy Me a Coffee" sponsor button (see [the GitHub docs](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/displaying-a-sponsor-button-in-your-repository) for more details on what this is and how to set it up)
- Bump `"node-version"` to `18` for GitHub Actions. This should reduce issues as we can internally target Node 18
- Bump package version to `"2.1.0"`
- Remove abbreviations in commands throughout for improved human readability/comprehension
- Remove `.npmignore` in favor of a concise `"files"` field in `package.json`
- Remove configuration for `*.js` files in `.editorconfig` to standardize 2-space indentation
- Remove `CHANGES.md` from `.gitignore` (this was something I used to use in my projects before discovering a better workflow for commit messages)
- Remove workspaces configuration from `package.json` in favor of a more traditional repository structure
- Update `LICENSE` to `LICENSE.md` from [IQAndreas/markdown-licenses](https://github.com/IQAndreas/markdown-licenses) for improved readability
- Update format of HTML in `README.md` for improved readability
- Update metadata (shields.io badges, etc.) in `README.md` to be consistent. Also adds a badge for license, version, gzip size, last commit, and buymeacoffee (removing the buymeacoffe button and nodei.co info)
- Update order of `package.json` fields to match the order in the [npm documentation](https://docs.npmjs.com/cli/configuring-npm/package-json)
- Update `package.json` to use abbreviated major version syntax (e.g. `"^1"` instead of `"^1.x.x"`)

## Testing

Please double-check and make sure these changes work. In an earlier version of this PR, I briefly tested the `npm pack` tarball output with a fresh bun (TypeScript) and npm project (JavaScript) and it seems to work fine.

1. Clone the repository
2. Run `npm install`
3. Run `npm pack`
4. Create a new project and run `npm install <path-to-tarball>`
5. See if types in the package works as expected (try with ts and js projects using require and import syntax)

## Package Size

The following demonstrates the reduction in unnecessary files to the npm package. Most notably, the package size has decreased to 13.0 kB (from 14.3 kB) and the unpacked size to 45.2 kB (from 65.9 kB)

<details open>
  <summary>v2.0.0</summary>

  ```
  📦  file-system-db@2.0.0
  === Tarball Contents === 
  225B   .editorconfig                 
  1.3kB  .github/workflows/publish.yaml
  549B   .github/workflows/test.yaml   
  11B    LICENSE                       
  6.7kB  README.md                     
  17.8kB lib/index.js                  
  11.6kB lib/LICENSE                   
  542B   lib/package.json              
  6.5kB  lib/README.md                 
  194B   lib/tsconfig.json             
  982B   package.json                  
  19.2kB tests/index.test.js           
  210B   tests/package.json            
  === Tarball Details === 
  name:          file-system-db                          
  version:       2.0.0                                   
  filename:      file-system-db-2.0.0.tgz                
  package size:  14.3 kB                                 
  unpacked size: 65.9 kB                                 
  shasum:        568554d4be7e77a2632aac4fe7177daa4d1734ae
  integrity:     sha512-mT7YmaseTCVXQ[...]fy7TlQwhMuIiA==
  total files:   13                                      
  ```

</details>
<details open>
  <summary>v2.1.0</summary>

  ```
    📦  file-system-db@2.1.0
  === Tarball Contents === 
  10.5kB LICENSE.md    
  6.6kB  README.md     
  9.8kB  index.d.ts    
  1.2kB  index.d.ts.map
  16.0kB index.js      
  992B   package.json  
  === Tarball Details === 
  name:          file-system-db                          
  version:       2.1.0                                   
  filename:      file-system-db-2.1.0.tgz                
  package size:  13.0 kB                                 
  unpacked size: 45.2 kB                                 
  shasum:        55367a1afd85144866ba763020deb5dda4b9d8b0
  integrity:     sha512-iXq9U+41o4OFq[...]8b/g7nTVJnEGg==
  total files:   6                                       
  ```

</details>
